### PR TITLE
GCW 2169 Appointment quotas

### PR DESCRIPTION
**Pull request by patrixr**
Adds 2 new tables

* Appointment Slot Presets
Presets define the default appointment quotas for each day of the week
* Appointment Slots
Each row defines a specific date and time, and the different appointment quotas for that slot.

Each controller has the default CRUD operations. 
The appointment slot controller has an added endpoint called /calendar, which aggregates the slots per date. When figuring out if there are open appointment slots for a certain date, appointment slots will be checked first. If nothing has been set, the presets will take effect.
Usage :
``` 
GET /calendar?from=2018-10-16&to=2018-10-31
```

### CODE REVIEW

File: app/controllers/api/v1/appointment_slots_controller.rb [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+      api :GET, '/v1/appointment_slots', "List upcoming appointment slots"
+      def index
+        render_with_timezone(@appointment_slots.upcoming.ascending)
+      end
+
+      api :GET, '/v1/appointment_slots/calendar', "List upcoming appointment slots aggregated by dates"
+      def calendar
+        from = params[:from] ? Date.parse(params[:from]) : Date.today
+        to = Date.parse(params[:to])
```
Comment (steveyken) [inline app/controllers/api/v1/appointment_slots_controller.rb]: Might be good to put an upper limit on the `to` field as this is a user entered param. Can avoid denial of service type attacks where someone enters to 2222-12-31 and we end up iterating every day between now and then in ```AppoinmentSlot.calendar```

File: app/controllers/api/v1/appointment_slots_controller.rb [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        render json: AppointmentSlot.calendar(from, to).to_json, status: 200
+      end
+
+      api :POST, "/v1/appointment_slots", "Add an appointment slot"
+      param_group :appointment_slot
+      def create
+        @appointment_slot.timestamp = @appointment_slot.timestamp.change(sec: 0)
```
Comment (steveyken) [inline app/controllers/api/v1/appointment_slots_controller.rb]: nice! I didn't know about ```change```

File: app/models/appointment_slot.rb [hunk 3]
```diff
@@ -0,0 +1,28 @@
+class AppointmentSlot < ActiveRecord::Base
+  validates :quota, numericality: { only_integer: true, greater_than_or_equal_to: 0 }
+
+  def self.calendar(from, to)
+    (from..to).map { |date| { date: date, slots: AppointmentSlot.for_date(date) } }
+  end
+
+  scope :upcoming, -> { where("timestamp >= ?", DateTime.now.beginning_of_day.to_s(:db)) }
```
Comment (steveyken) [inline app/models/appointment_slot.rb]: do you need a utc conversion here? i.e. ``` DateTime.now.beginning_of_day.utc.to_s(:db)```

File: app/models/appointment_slot_preset.rb [hunk 4]
```diff
@@ -0,0 +1,8 @@
+class AppointmentSlotPreset < ActiveRecord::Base
+  validates :day, numericality: { only_integer: true, greater_than: 0, less_than: 8  }
```
Comment (steveyken) [inline app/models/appointment_slot_preset.rb]: nice!

File: spec/controllers/api/v1/appointment_slot_presets_controller_spec.rb [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+      end
+
+      it 'returns all preset slots' do
+        5.times { FactoryBot.create :appointment_slot_preset }        
+        get :index
+        expect(parsed_body['appointment_slot_presets'].count).to eq(AppointmentSlotPreset.count)
+      end
+    end
+
+    context 'When logged in without any rights' do
```
Comment (steveyken) [inline spec/controllers/api/v1/appointment_slot_presets_controller_spec.rb]: Good edge cases to be sure!

File: spec/controllers/api/v1/appointment_slot_presets_controller_spec.rb [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+        expect(parsed_body['appointment_slot_preset']['day']).to eq(payload['day'])
+      end
+
+      it "should update an appointment slot preset if it exists already" do
+        existing_preset = FactoryBot.create :appointment_slot_preset, day: 1, hours: 14, minutes: 30, quota: 3
+        expect(AppointmentSlotPreset.count).to eq(1)
+
+        post :create, appointment_slot_preset: { day: 1, hours: 14, minutes: 30, quota: 10 }
+        expect(response.status).to eq(201)
+        expect(AppointmentSlotPreset.count).to eq(1)
```
Comment (steveyken) [inline spec/controllers/api/v1/appointment_slot_presets_controller_spec.rb]: good test!

File: spec/controllers/api/v1/appointment_slots_controller_spec.rb [hunk 7]
```diff
@@ -0,0 +1,151 @@
+require 'rails_helper'
+
+RSpec.describe Api::V1::AppointmentSlotsController, type: :controller do
+  let(:order_administrator) { create(:user, :order_administrator, :with_can_manage_settings )}
+  let(:no_permission_user) { create :user }
+  let(:parsed_body) { JSON.parse(response.body) }
+
+  def now
+    DateTime.now.change(sec: 0, usec: 0)
```
Comment (steveyken) [inline spec/controllers/api/v1/appointment_slots_controller_spec.rb]: Beware of the Time Lords!

image

File: spec/factories/sequences.rb [hunk 8]
```diff
@@ -65,7 +65,9 @@
   end
 
   sequence :permissions_roles do |n|
-    @permissions_roles ||= YAML.load_file("#{Rails.root}/db/permissions_roles.yml")
+    roles = YAML.load_file("#{Rails.root}/db/permissions_roles.yml")
```
Comment (steveyken) [inline spec/factories/sequences.rb]: Good catch

File: app/models/appointment_slot.rb [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+
+  scope :ascending, -> { order('timestamp ASC').order('quota ASC') }
+
+  def self.for_date(date)
```
Comment (steveyken) [inline app/models/appointment_slot.rb]: This is nice!

File: app/controllers/api/v1/appointment_slots_controller.rb (hunk 1 above)
Comment (patrixr) [reply app/controllers/api/v1/appointment_slots_controller.rb]: good point !

File: app/controllers/api/v1/appointment_slots_controller.rb [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+        if AppointmentSlot.find_by(timestamp: @appointment_slot.timestamp)
+          render_error('Timeslot already exists')
```
[the remark this answers is not in the document]
Comment (patrixr) [reply app/controllers/api/v1/appointment_slots_controller.rb]: will do

File: app/controllers/api/v1/appointment_slots_controller.rb (hunk 2 above)
Comment (patrixr) [reply app/controllers/api/v1/appointment_slots_controller.rb]: Yup. Had issues in comparing timestamps. So I'm making sure seconds aren't saved

File: app/models/ability.rb [hunk 11]
```diff
@@ -80,6 +81,12 @@ def address_abilities
     can [:create, :show, :destroy], Address, addressable_type: "Contact" if can_manage_delivery_address?
   end
 
+  def appointment_slot_abilities
+    return unless can_manage_settings?
+    can [:create, :index, :destroy, :update], AppointmentSlotPreset
+    can [:create, :index, :destroy, :update, :calendar], AppointmentSlot
```
[the remark this answers is not in the document]
Comment (patrixr) [reply app/models/ability.rb]: @steveyken Calendar already exists. But I will add a field to the response which specifies if the day is full or not

File: app/models/appointment_slot.rb (hunk 3 above)
Comment (patrixr) [reply app/models/appointment_slot.rb]: good point

File: spec/controllers/api/v1/appointment_slots_controller_spec.rb (hunk 7 above)
Comment (patrixr) [reply spec/controllers/api/v1/appointment_slots_controller_spec.rb]: image

**Pull request merged**